snuba/query/snql/__init__.py

```diff
@@ -1,23 +1,280 @@
+from enum import Enum
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    List,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+)
+
 from parsimonious.grammar import Grammar
+from parsimonious.nodes import Node, NodeVisitor
+
 from snuba.datasets.dataset import Dataset
-from snuba.query.logical import Query
+from snuba.query.dsl import divide, minus, multiply, plus
+from snuba.query.expressions import (
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    FunctionCall,
+    Literal,
+)
+from snuba.query.logical import Query, SelectedExpression
+
+
+snql_grammar = Grammar(
+    fr"""
+    query_exp             = match_clause? where_clause? collect_clause? having_clause? order_by_clause?
+    match_clause          = space* "MATCH" space* "(" clause ")" space*
+    where_clause          = space* "WHERE" clause space*
+    collect_clause        = space* "COLLECT" collect_list space*
+    order_by_clause       = space* "ORDER BY" collect_list ("ASC"/"DESC") space*
+
+    collect_list          = collect_columns* (low_pri_arithmetic)
+    collect_columns       = low_pri_arithmetic space* comma space*
 
+    having_clause         = space* "HAVING" clause space*
+    clause                = space* ~r"[-=><\w]+" space*
 
-grammar = Grammar(
-    r"""
-    query_exp = match_clause where_clause? collect_clause? having_clause? order_by_clause?
-    match_clause = space* "MATCH" "(" clause ")" space*
-    where_clause = space* "WHERE" clause space*
-    collect_clause = space* "COLLECT" clause "BY" clause space*
-    having_clause = space* "HAVING" clause space*
-    order_by_clause = space* "ORDER BY" clause ("ASC"/"DESC") space*
+    low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
+    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    low_pri_tuple         = low_pri_op high_pri_arithmetic
+    high_pri_tuple        = high_pri_op arithmetic_term
+    arithmetic_term       = (space*) (function_call / numeric_literal / column_name) (space*)
+    low_pri_op            = "+" / "-"
+    high_pri_op           = "/" / "*"
+    param_expression      = low_pri_arithmetic / quoted_literal
+    parameters_list       = parameter* (param_expression)
+    parameter             = param_expression space* comma space*
+    function_call         = function_name open_paren parameters_list? close_paren (open_paren parameters_list? close_paren)?
+    simple_term           = quoted_literal / numeric_literal / column_name
+    literal               = ~r"[a-zA-Z0-9_\.:-]+"
+    quoted_literal        = "'" string_literal "'"
+    string_literal        = ~r"[a-zA-Z0-9_\.\+\*\/:-]*"
+    numeric_literal       = ~r"-?[0-9]+(\.[0-9]+)?(e[\+\-][0-9]+)?"
+    column_name           = ~r"[a-zA-Z_][a-zA-Z0-9_\.]*"
+    function_name         = ~r"[a-zA-Z_][a-zA-Z0-9_]*"
+    open_paren            = "("
+    close_paren           = ")"
+    space                 = " "
```

Comment:
This part is the same we had in the clickhouse parser in the old language. Is that correct or did you make changes?